src/us/kbase/workspace/database/provenance/ResolvablePID.java

```diff
@@ -0,0 +1,118 @@
+package us.kbase.workspace.database.provenance;
+
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * A resolvable permanent identifier for a resource.
+ */
+public class ResolvablePID {
+
+	private final String id;
+	private final String description;
+	private final String relationshipType;
+
+	private ResolvablePID(final String id, final String description, final String relationshipType) {
+		this.id = id;
+		this.relationshipType = relationshipType;
+		this.description = description;
+	}
+
+	/**
+	 * Get the ID, for example DOI:10.25982/59912.37.
+	 * 
+	 * @return the id.
+	 */
+	public String getId() {
+		return id;
+	}
+
+	/**
+	 * Get a free text description of the ID.
+	 * 
+	 * @return the description, if present.
+	 */
+	public Optional<String> getDescription() {
+		return Optional.ofNullable(description);
+	}
+
+	/**
+	 * Get the relationship between the ID and the resource, for example
+	 * "isSupplementTo" or "isNewVersionOf".
```

Comment:
Hmm, I don't want to open a can of worms, but should this be an Enum or some other type of controlled vocab?

Reply:
yes, it's coming from a controlled vocab. I'm in two minds about what to do here (and for other fields that would ideally be enums) -- whether it would be better to validate in the credit engine (where I have tools that do all the work for me) and have the WS just save the data, or whether the WS should also validate enums such as these too. Enums are the the parts of the schema that I think are most likely to need editing in the future, and it will be a nuisance if the WS has to be updated each time.

A few other possible ideas I've thought of: 
- build a minimal parser for the schema that could pull in the appropriate enum values for the appropriate classes;
- give the WS a copy of the schema in JSONschema format, validate the incoming data according to the schema (using an existing Java validator), so errors around enums, etc., get caught there and these classes don't need to do much validation beyond type checking; 
- have the WS launch a subprocess that runs the validator (python library);
- extend the existing linkml Java class generator code to be able to create fully-fleshed out Java objects from the schema.

For now, since the credit engine is the only likely source of credit metadata, I feel it's reasonable to go for a simple implementation where the WS doesn't check enums, but I'm not sure that's a good long term solution.

Reply:
(we did discuss some of this before, with the conclusion being KISS for the time being; however, I'm not sure I mentioned the other strategies that I thought of to deal with the issue)

Reply:
Since only admins will be able to add credit info for now, validating outside the WS seems fine, with documentation somewhere in a relevant place in the WS repo warning future coders that they made need to do more validation if the method is opened up.

For the future, I think giving the workspace the JSONschmea seems reasonable, or it probably wouldn't be too hard to make a script that extracts the relationship values from the schema and creates a java Enum (or is that what you mean by #1?).

Reply:
> it probably wouldn't be too hard to make a script that extracts the relationship values from the schema and creates a java Enum (or is that what you mean by https://github.com/kbase/workspace_deluxe/issues/1?).

Yep.

I have been looking at JSONschema tools this morning and there are some Java generators that can produce Java classes (of various different kinds / styles) from a JSONschema, so future enhancements might include adding JSONschema validation to the WS and implementing the Java version of the schema using one of those generators.